src/rules/declaration-block-no-redundant-longhand-properties/index.js

```diff
@@ -0,0 +1,85 @@
+import {
+  isEqual,
+  isString,
+  transform,
+} from "lodash"
+import {
+  optionsMatches,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+import shorthandData from "../../reference/shorthandData"
+
+export const ruleName = "declaration-block-no-redundant-longhand-properties"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (props) => (
+    `Expected longhand properties "${props}"`
+  ),
+})
+
+const customProps = {
+  "border-color": "border-*-color",
+  "border-style": "border-*-style",
+  "border-width": "border-*-width",
+  "border": "border-*-*",
+}
+
+export default function (actual, options) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, { actual }, {
+      actual: options,
+      possible: {
+        ignoreShorthands: [isString],
+      },
+      optional: true,
+    })
+    if (!validOptions) { return }
+
+    const longhandProperties = transform(shorthandData, (result, values, key) => {
+      values.forEach((value) => {
+        (result[value] || (result[value] = [])).push(key)
+      })
+    })
+
+    root.walkRules(check)
+    root.walkAtRules(check)
+
+    function check(statement) {
+      const longhandDeclarations = {}
+      // Shallow iteration so nesting doesn't produce
+      // false positives
+      statement.each(node => {
+        if (node.type !== "decl") { return }
+
+        const prop = node.prop.toLowerCase()
+
+        if (optionsMatches(options, "ignoreShorthands", prop)) { return }
```

Comment:
We don't need to ignore shorthands *here* — we need to ignore them when the rule would otherwise ask you to use them.